src/node-file-trace.ts

```diff
@@ -12,6 +12,30 @@ const fsReadFile = fs.promises.readFile;
 const fsReadlink = fs.promises.readlink;
 const fsStat = fs.promises.stat;
 
+type ParseSpecifierResult = {
+  path: string;
+  queryString: string | null
+}
+
+// Splits an ESM specifier into path and querystring (including the leading `?`). (If the specifier is CJS,
+// it is passed through untouched.)
+export function parseSpecifier(specifier: string, cjsResolve: boolean = true): ParseSpecifierResult {
+  let path = specifier;
+  let queryString = null;
+
+  if (!cjsResolve) {
+    // Regex which splits a specifier into path and querystring, inspired by that in `enhanced-resolve`
+    // https://github.com/webpack/enhanced-resolve/blob/157ed9bcc381857d979e56d2f20a5a17c6362bff/lib/util/identifier.js#L8
+    const match = /^(#?(?:\0.|[^?#\0])*)(\?(?:\0.|[^\0])*)?$/.exec(specifier);
```

Comment:
This looks slightly different from the webpack regex. 

Can we instead use `url.parse()` or `new URL()` to safely parse?

Reply:
> This looks slightly different from the webpack regex.
>
> Can we instead use url.parse() or new URL() to safely parse?

In my testing, node itself seemed to treat `import * as something from './someFile.js?things#stuff`, `import * as something from './someFile.js?things`, and `import * as something from './someFile.js` the same way, namely that it stripped the `?` and everything after it off before trying to resolve the file. I therefore wanted to do the same thing. Before landing on this regex (which is the original regex, minus treating the fragment as a separate entity - see the screenshot below), I considered:

- Using `indexOf` on `'?'` and then using `slice` to cut the string off at that point. Straightforward, but brute-force-y, and I've had brute force bite me enough times when handling edge cases I hadn't thought of that I wanted to out-source it to something purpose-built and presumably battle-tested.

- Using `url.parse`. This splits the querystring and fragment off into separate parts I'd just have to put back together again. Also, my linter got mad because [it's deprecated](https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost).

- Using `new URL()`. This similarly splits the querystring and fragment off separately. It also errors out when given a path without `http://` (or similar) at the beginning.

None of them seemed ideal, and so I decided to see how webpack does it (since I was in the nextjs nft webpack plugin with the debugger already), figuring that the most likely context for someone having the querystring problem in the first place was them using a webpack loader with options. After a lot of stepping through code, eventually I landed in the `enhanced-resolve` file I linked to in the comment. As I mentioned above, you're right that it's been changed, but only to combine the querystring and fragment into a single group.

![image](https://user-images.githubusercontent.com/14812505/204605370-0bf11d19-abb3-4eff-94c6-f85d0f10ae3c.png)

(I gave the regex I'm using above a name the same length as the original regex's name in `enhanced-resolve` so it's easier to compare.)

I can certainly make `url.parse` or `new URL()` work if you'd like, but given the above IMHO the regex is still the best option. WDYT?

Reply:
`url.parse()` is not really deprecated because its the only safe way to parse a relative url since `new URL()` only works for absolute urls. The docs mention the deprecation was revoked 
<img width="717" alt="image" src="https://user-images.githubusercontent.com/229881/208211111-829021ba-92e5-4760-badd-beaf8d2d328c.png">

Also, I trust node's own url parsing more than a modified regex which might suffer from [ReDoS](https://en.wikipedia.org/wiki/ReDoS).

Now that there are tests, we should be able to swap this implementation between regex, url.parse() or new URL() easily.